README.rst
=======================
README for the fred2pdf
=======================


Installation
============

Installing fred2pdf is a simple process. Just execute the setup.py script:

``$ sudo python setup.py install``

Please note that the installation has to be performed using 
the administrators account (i.e. root on linux). Installation copy files
into standard python modules folder and creates default configuration file.
The default templates are copied into folder /usr/share/fred2pdf_docs/

Dependences
-----------

fred2pdf requires these packages:

* Python
* PIL (Python Imaging Library)
* Reportlab
* Tiny RML2PDF

Fonts
-----

fred2pdf needs also truetype fonts. Installation looked for them in the path
``/usr/share/fonts``


Running
=======

Just call script doc2pdf.py. This script assumes RML template file and print PDF into
standard output. The RML file you obtain from transformation the XSL template 
with XML data file. Use some XSLT parser (i.e. xsltproc).
You can join whole process into one pipe:

``$ xsltproc /usr/fred2pdf_docs/invoice.xsl /usr/fred2pdf_docs/invoice.xml | doc2pdf.py > invoice.pdf``


Translations
------------

Translations of the templates are available by option lang.

Example of the english translation (default is cs):

``$ xsltproc -stringparam lang en /usr/fred2pdf_docs/invoice.xsl /usr/fred2pdf_docs/invoice.xml | doc2pdf.py > invoice.pdf``


ChangeLog
2020-08-05 (2.22.3)

    * Fix layout error

2020-04-17 (2.22.2)

    * Update helios template for new invoice prefixes

2020-01-31 (2.22.1)

    * Fix rpm for RHEL8 and F31

2020-01-08 (2.22.0)

    * Add administrative fee item code to invoice template

2019-11-20 (2.21.1)

    * Update spec file for F31 and Centos/RHEL 8

2019-09-11 (2.21.0)

    * Add monthly fee item code to invoice template

    * Add note about VAT calculation

2019-07-15 (2.20.0)

    * Record statement fixes and enhancements
        - date formatting
        - trim organization field
        - footer text removed

2019-03-18 (2.19.0)

    * License GNU GPLv3+

2019-01-16 (2.18.1)

    * Fix - remove 'X X X' line from mojeid letter

2018-09-05 (2.18.0)

    * Font definitions moved to templates

    * Use setuptools for distribution

2018-08-14 (2.17.1)

    * Helios template fixes

2018-04-23 (2.17.0)

    * Add new template for personal info public request

2018-03-27 (2.16.1)

    * Fix registry record statement template - timezone

2018-02-14 (2.16.0)

    * Enhancements of contact verification pdf (letter)
        - Add QR code with url
        - Font size changes
        - Text and grammar fixes
        - Removed duplicate footer

2017-09-12 (2.15.0)

    * Add registry record statement templates

2017-10-11 (2.14.1)

    * Update text of mojeid validation pdf
        - Remove second paragraph about validation by certificate
        - Show paragraph with "Datova schranka" only to organisations
        - Fix link to the site mojeid/validation

2017-03-02 (2.14.0)

    * Remove Arial font dependency

2017-05-03 (2.13.3)

    * Fix invoice template (appendix padding)

2017-03-09 (2.13.2)

    * Fedora packaging

2016-12-20 (2.13.1)

    * Fix text of mojeid validation pdf

2016-11-29 Ales Friedl, Jiri Sadek (2.13.0)

    * Remove fax in cz.nic design template, use mobile phone instead

    * Fix examples

    * Remove generated files

2016-09-07 Zdenek Bohm, Zuzana Ansorgova (2.12.0)

    * configuration file documentation

    * compatibility with reportlab 3.3.0

2016-06-13 Jiri Sadek (2.11.3)

    * mojeid validation pdf (datova schranka)

2016-06-07 Jaromir Talir (2.11.2)

    * new helios version template fixes

2016-03-22 Michal Strnad, Jaromir Talir (2.11.1)

    * Fix rpm build

    * Fix example data

2015-05-19 Michal Strnad (2.11.0)

    * new mojeid card letter template

2015-01-27 Michal Strnad, Jan Korous (2.10.0)

    * new (improved) content of admin. verification letter

    * new design of mojeid validation pdf

    * fix invoice typo

    * removed company_name field from shared address template

    * fix condition for domestic letters (country=czech republic)

2014-12-31 Jan Zima (2.9.1)

    * company address change

2014-10-17 Michal Strnad (2.9.0)

    * new mojeid re-identification letter

    * mojeid validation pdf - address type defined in text

    * mojeid pin3 letters fixes (QR code)

2014-09-17 Jaromir Talir (2.8.2)

    * helios template update

2014-08-29 Michal Strnad (2.8.1)

    * fix template for mojeid pin3 letter (address country)

2014-08-01 Michal Strnad, Zdenek Bohm (2.8.0)

    * new template for new mojeid pin3 letter

    * fix admin. contact verification letter long address bug

2014-06-12 Zdenek Bohm (2.7.0)

    * new templates for admin. contact verification letters

    * several grammar fixes accross all templates

    * change of association registration note in all templates

2014-02-18 Jaromir Talir (2.6.3)

    * Fix trml2pdf path checking in setup

2014-02-13 Michal Strnad (2.6.2)

    * Fix mojeid validation template (id-card copy statement, add new validation places)

2014-02-05 Zdenek Bohm (2.6.1)

    * Address position fix (verification, domain expiration)

2014-01-07 Zdenek Bohm (2.6.0)

    * New design

2013-08-22 Zdenek Bohm, Jiri Sadek (2.5.3)

    * Fix pdf fonts rendering (accented characters), should work with reportlab 2.4, 2.5, 2.6

2013-06-07 Zdenek Bohm, Vlastimil Zima (2.5.2)

    * Fix error message when font file was not found

    * Updating setup.cfg and setup.py according recent fred-distutils changes

    * New CZ.NIC logos

2012-11-19 Jaromir Talir (2.5.1)

    * mojeid pin3 letter template changes

2012-09-05 Juraj Vicenik, Jan Zima (2.5.0)

    * contact verification letter templates

    * mojeid_auth tag changed to contact_auth in mojeid letter template

    * fix mojeid validation template

2012-05-15 Zdenek Bohm (2.4.0)

    * invoice template fix (advance payments - bold text)

    * new invoice examples

2012-04-12 Zdenek Bohm (2.3.5)

    * invoice template fix (summarize item line)

2012-03-26 Zdenek Bohm (2.3.4)

    * mojeid templates fixes

2012-03-14 Zdenek Bohm, Juraj Vicenik (2.3.3)

    * mojeid templates changes

    * letter address format fix (stateorprovince)

2011-11-04 Zdenek Bohm (2.3.2)

    * account invoice fixes (translations)

2011-11-02 Zdenek Bohm, Jaromir Talir (2.3.1)

    * fix rpm build

    * account invoice fixes (service codes)

2011-10-18 Zdenek Bohm, Jaromir Talir, Jan Zima (2.3.0)

    * updated FredTable constructor default parameters (to be compatible with
      tinyerp-server-4.2.3.4-7)

    * helios template update

    * account invoice template update
        - new template parameters
        - formatting changes
        - translations updates

2011-06-02 Vit Vomacko (2.2.0)

    * removed local freddist

2010-12-13 Juraj Vicenik (2.1.7)

    * If country is Czech republic, don't write it to the address

    * Changed office hours in mojeID validation document

2010-11-04 Juraj Vicenik (2.1.6)

    * Modified documents for MojeID - validation

    * Changed XML format for validation - handle is allowed

2010-10-25 Juraj Vicenik (2.1.5)

    * Fixes in mojeid letters

2010-10-23 Jaromir Talir , Juraj Vicenik (2.1.4)

    * Fixes in mojeid letters

2010-10-18 Zdenek Bohm, Juraj Vicenik (2.1.3)

    * New documents for Mojeid (identification and validation letters)


freddist/command/install_data.py
import types, os, re, sys
from distutils import util
from distutils.command.install_data import install_data as _install_data
from install_parent import install_parent

# freddist install_data came with one enhancement. It regards system directories.
# And first simple example. This is part of core.setup function:
#
# 01     setup(name='some_name',
# 02             author='your name',
# 03             #important part goes now
# 04             data_files = [
# 05                 ('SYSCONFDIR/some_name',
# 06                     [
# 07                         'file_no1.conf',
# 08                         'file_no2.conf'
# 09                     ]
# 10                 ),
# 11                 ('DOCDIR',
# 12                     [
# 13                         'document_no1.html'
# 14                     ]
# 15                 ),
# 16                 ('LOCALSTATEDIR/whatever', ),
# 17             ],
# 18             another setup stuff....,
#
# As you can see, in data_files is used some strange `SYSCONFDIR' and `DOCDIR'
# option. These are during install phase replaced by fully expanded directory
# names. For example `file_no1.conf' and `file_no2.conf' will be installed into
# `PREFIX/etc/some_name' directory. And since `PREFIX' is normally `/usr/local'
# then full path to first file will be `/usr/local/etc/some_name/file_no1.conf'.
# All these setting can be overriden by proper options.
# On line 16 is example of creating empty directory.

class install_data(_install_data, install_parent):
    user_options = _install_data.user_options
    user_options.append(('prefix=', None,
        'installation prefix'))
    user_options.append(('bindir=', None,
        'user executables [PREFIX/bin]'))
    user_options.append(('sbindir=', None,
        'system admin executables [PREFIX/sbin]'))
    user_options.append(('sysconfdir=', None, 
        'System configuration directory [PREFIX/etc]'))
    user_options.append(('libexecdir=', None,
        'Program executables [PREFIX/libexec]'))
    user_options.append(('localstatedir=', None,
        'Modifiable single machine data [PREFIX/var]'))
    user_options.append(('libdir=', None,
        'object code libraries [PREFIX/lib]'))
    user_options.append(('pythondir=', None,
        'python directory [LIBDIR/python%d.%d]' %
        (sys.version_info[0], sys.version_info[1])))
    user_options.append(('purelibdir=', None,
        'python pure libraries [LIBDIR/python%d.%d/site-packages]' %
        (sys.version_info[0], sys.version_info[1])))
    user_options.append(('datarootdir=', None,
        'read only architecture-independent data root [PREFIX/share]'))
    user_options.append(('datadir=', None,
        'read only architecture-independent data [DATAROOTDIR]'))
    user_options.append(('infodir=', None,
        'info documentation [DATAROOTDIR/info]'))
    user_options.append(('mandir=', None,
        'man documentation [DATAROOTDIR/man]'))
    user_options.append(('docdir=', None,
        'documentation root [DATAROOTDIR/doc/NAME]'))
    user_options.append(('localedir=', None,
        'locale-dependent data [DATAROOTDIR/locale]'))
    user_options.append(('preservepath', None, 
        'Preserve path(s) in configuration file(s).'))
    user_options.append(('dont-record', None,
        'do not record list of installed files'))
    user_options.append(('dont-create-pycpyo', None,
        'do not create compiled pyc and optimized pyo files'))
    user_options.append(('no-check-deps', None,
        'do not check dependencies'))

    # directory patterns which install_data recognize
    dir_patts = ['PREFIX', 'SYSCONFDIR', 'LOCALSTATEDIR', 'LIBEXECDIR',
            'LIBDIR', 'DATAROOTDIR', 'DATADIR', 'MANDIR', 'DOCDIR',
            'INFODIR', 'SBINDIR', 'BINDIR', 'LOCALEDIR', 'PYTHONDIR',
            'PURELIBDIR']

    boolean_options = _install_data.boolean_options
    boolean_options.append('preservepath')
    boolean_options.append('dont_record')
    boolean_options.append('dont_create_pycpyo')
    boolean_options.append('no-check-deps')

    def __init__(self, *attrs):
        _install_data.__init__(self, *attrs)
        install_parent.__init__(self, *attrs)

    def replaceSpecialDir(self, dir):
        """
        Method purpose is to replace `special directory' pattern passed to
        freddist.core.setup function in its data_files list. For example
        data_files could look like this:
            data_files = [('SYSCONFDIR/some_dir', ['file_no1', 'file_no2'])]
        In this example `file_no1' and `file_no2' will be copied into
        `SYSCONFDIR/some_dir' directory where `SYSCONFDIR' will be replaced
        with actual setting (by default it is `prefix/etc' where prefix is
        `/usr/local'. So whole path will be `/usr/local/etc/some_dir').
        Valid patterns are emplaced in self.`dir_patts' variable.
        """
        for str in self.dir_patts:
            s = re.search(str, dir)
            if s:
                dir = self.__dict__[str.lower()] + dir[s.end():]
        return dir

    def initialize_options(self):
        _install_data.initialize_options(self)
        install_parent.initialize_options(self)
        self.prefix = None
        self.root = None
        self.record = None

    def finalize_options(self):
        _install_data.finalize_options(self)
        self.set_undefined_options('install',
                ('prefix', 'prefix'),
                ('sysconfdir', 'sysconfdir'),
                ('localstatedir', 'localstatedir'),
                ('libexecdir', 'libexecdir'),
                ('preservepath', 'preservepath'),
                ('root', 'root'),
                ('libdir', 'libdir'),
                ('datarootdir', 'datarootdir'),
                ('datadir', 'datadir'),
                ('mandir', 'mandir'),
                ('docdir', 'docdir'),
                ('infodir', 'infodir'),
                ('bindir', 'bindir'),
                ('sbindir', 'sbindir'),
                ('localedir', 'localedir'),
                ('pythondir', 'pythondir'),
                ('purelibdir', 'purelibdir'),
                ('dont_create_pycpyo', 'dont_create_pycpyo'),
                ('dont_record', 'dont_record'))
        install_parent.finalize_options(self)
        if not self.record and not self.dont_record:
            self.record = 'install.log'


    def run(self):
        #FREDDIST line added
        self.mkpath(self.install_dir)
        for f in self.data_files:
            if type(f) is types.StringType:
                #FREDDIST next line changed
                if not os.path.exists(f):
                    f = util.convert_path(os.path.join(self.srcdir, f))
                if self.warn_dir:
                    self.warn("setup script did not provide a directory for "
                              "'%s' -- installing right in '%s'" %
                              (f, self.install_dir))
                # it's a simple file, so copy it
                (out, _) = self.copy_file(f, self.install_dir)
                self.outfiles.append(out)

                if out.endswith('.py') and not self.dont_create_pycpyo:
                    os.system('python -c "import py_compile; \
                            py_compile.compile(\'%s\')"' % out)
                    self.outfiles.append(out)
                    print "creating compiled %s" % out + 'c'
                    os.system('python -O -c "import py_compile; \
                            py_compile.compile(\'%s\')"' % out)
                    self.outfiles.append(out)
                    print "creating optimized %s" % out + 'o'
            else:
                # it's a tuple with path to install to and a list of files
                dir = util.convert_path(self.replaceSpecialDir(f[0]))
                if not os.path.isabs(dir):
                    dir = os.path.join(self.install_dir, dir)
                elif self.root:
                    dir = util.change_root(self.root, dir)
                self.mkpath(dir)

                if len(f) == 1 or f[1] == []:
                    # If there are no files listed, the user must be
                    # trying to create an empty directory, so add the
                    # directory to the list of output files.
                    self.outfiles.append(dir)
                    print "creating directory %s" % dir
                else:
                    # Copy files, adding them to the list of output files.
                    for data in f[1]:
                        #FREDDIST next line changed
                        if not os.path.exists(data):
                            data = util.convert_path(
                                    os.path.join(self.srcdir, data))
                        (out, _) = self.copy_file(data, dir)
                        self.outfiles.append(out)
                        if 'bin' in out.split(os.path.sep) or 'sbin' in out.split(os.path.sep):
                            os.chmod(out, 0755)

                        if out.endswith('.py') and not self.dont_create_pycpyo:
                            os.system('python -c "import py_compile; \
                                    py_compile.compile(\'%s\')"' % out)
                            self.outfiles.append(out + 'c')
                            print "creating compiled %s" % out + 'c'
                            os.system('python -O -c "import py_compile; \
                                    py_compile.compile(\'%s\')"' % out)
                            self.outfiles.append(out + 'o')
                            print "creating optimized %s" % out + 'o'
